compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -874,7 +874,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let found = self.resolve_vars_with_obligations(found);
 
         let in_loop = self.is_loop(id)
-            || self.tcx.hir().parent_iter(id).any(|(parent_id, _)| self.is_loop(parent_id));
+            || self
+                .tcx
+                .hir()
+                .parent_iter(id)
+                .take_while(|(_, node)| {
+                    !matches!(
+                        node,
+                        hir::Node::Item(_)
+                            | hir::Node::Expr(hir::Expr { kind: hir::ExprKind::Closure(_), .. })
```

Comment:
Just noticed that we should probably also check for `async` blocks.

Reply:
We lower async to closures before we get to this point.